File: python/graphmind/transport.py
"""WebSocket transport: lazy connect, handshake, background retry, fail-open.

Port of ``packages/client/src/transport.ts``. Everything here runs on the
shared :mod:`graphmind.runtime` loop thread; the only cross-thread entry points
are :meth:`Transport.start`, :meth:`Transport.kick`, :meth:`Transport.enqueue`
and :meth:`Transport.dispose`, all of which are non-blocking.

The transport never raises into callers: every failure degrades to "detached"
plus a rate-limited warning, and a background retry keeps trying every
``retry_interval`` seconds.
"""

from __future__ import annotations

import asyncio
from collections import deque
from typing import Any, Callable, Deque, Dict, Optional

from .protocol import PROTOCOL_VERSION, parse_envelope_json
from .runtime import runtime
from .safe import RateLimitedWarner

#: Hard cap on frames waiting for a slow socket. Older frames are dropped
#: first; the ring buffer is what guarantees replay-on-attach, not this queue.
MAX_OUTBOX = 10_000


def _resolve_connect() -> Optional[Callable[..., Any]]:
    """The websockets client factory, tolerating old and new package layouts."""
    try:
        from websockets.asyncio.client import connect  # websockets >= 13

        return connect
    except Exception:
        pass
    try:  # pragma: no cover - only on ancient websockets
        from websockets.client import connect as legacy_connect

        return legacy_connect
    except Exception:
        return None


class TransportHooks:
    """Callbacks the session installs. All are invoked on the loop thread."""

    __slots__ = ("build_hello", "on_attached", "on_detached", "on_control")

    def __init__(
        self,
        build_hello: Callable[[], str],
        on_attached: Callable[[Dict[str, Any]], None],
        on_detached: Callable[[], None],
        on_control: Callable[[Dict[str, Any]], None],
    ) -> None:
        self.build_hello = build_hello
        self.on_attached = on_attached
        self.on_detached = on_detached
        self.on_control = on_control


class Transport:
    def __init__(
        self,
        url: str,
        hooks: TransportHooks,
        warner: RateLimitedWarner,
        connect_timeout: float = 0.3,
        handshake_timeout: float = 1.0,
        retry_interval: float = 10.0,
    ) -> None:
        self._url = url
        self._hooks = hooks
        self._warner = warner
        self._connect_timeout = connect_timeout
        self._handshake_timeout = handshake_timeout
        self._retry_interval = retry_interval

        self._state = "idle"  # idle | connecting | attached | disposed
        self._started = False
        self._outbox: Deque[str] = deque()
        self._wake: Optional[asyncio.Event] = None
        self._kick: Optional[asyncio.Event] = None
        self._supervisor: Optional["asyncio.Task[None]"] = None
        self._connect = _resolve_connect()

    # -- public, thread-safe --------------------------------------------------

    @property
    def attached(self) -> bool:
        return self._state == "attached"

    @property
    def state(self) -> str:
        return self._state

    def start(self) -> None:
        """Begin connecting (idempotent). Called lazily on first session use."""
        if self._started or self._state == "disposed":
            return
        self._started = True
        if self._connect is None:
            self._warner.warn(
                "transport-no-impl",
                "the `websockets` package is not importable; GraphMind stays detached "
                "(pip install graphmind-ai[all] or pip install websockets)",
            )
            return
        runtime.call_soon(self._start_supervisor)

    def kick(self) -> None:
        """Force an attempt now instead of waiting out the retry interval."""
        if self._state == "disposed":
            return
        if not self._started:
            self.start()
            return
        runtime.call_soon(self._set_kick)

    def enqueue(self, frame: str) -> None:
        """Queue one serialized frame for sending. Non-blocking, any thread."""
        if self._state != "attached":
            return
        outbox = self._outbox
        outbox.append(frame)
        while len(outbox) > MAX_OUTBOX:
            try:
                outbox.popleft()
            except IndexError:  # pragma: no cover - concurrent drain
                break
        runtime.call_soon(self._set_wake)

    def dispose(self) -> None:
        if self._state == "disposed":
            return
        self._state = "disposed"
        self._outbox.clear()
        runtime.call_soon(self._cancel_supervisor)

    # -- loop-thread internals ------------------------------------------------

    def _set_wake(self) -> None:
        if self._wake is not None:
            self._wake.set()

    def _set_kick(self) -> None:
        if self._kick is not None:
            self._kick.set()

    def _cancel_supervisor(self) -> None:
        task = self._supervisor
        self._supervisor = None
        if task is not None and not task.done():
            task.cancel()

    def _start_supervisor(self) -> None:
        if self._state == "disposed" or self._supervisor is not None:
            return
        self._wake = asyncio.Event()
        self._kick = asyncio.Event()
        loop = asyncio.get_event_loop()
        self._supervisor = loop.create_task(self._supervise())

    async def _supervise(self) -> None:
        while self._state != "disposed":
            if self._kick is not None:
                self._kick.clear()
            try:
                await self._attempt()
            except asyncio.CancelledError:
                raise
            except Exception as exc:  # pragma: no cover - defence in depth
                self._warner.warn("transport-loop", "transport loop error", exc)
            if self._state == "disposed":
                return
            try:
                if self._kick is None:
                    await asyncio.sleep(self._retry_interval)
                else:
                    await asyncio.wait_for(self._kick.wait(), timeout=self._retry_interval)
            except asyncio.TimeoutError:
                pass
            except asyncio.CancelledError:
                raise
            except Exception:  # pragma: no cover
                await asyncio.sleep(self._retry_interval)

    async def _attempt(self) -> None:
        if self._connect is None or self._state == "disposed":
            return
        self._state = "connecting"
        try:
            connection = await asyncio.wait_for(
                self._connect(self._url, open_timeout=None, max_queue=64),
                timeout=self._connect_timeout,
            )
        except asyncio.CancelledError:
            self._state = "idle"
            raise
        except asyncio.TimeoutError:
            self._state = "idle"
            self._warner.warn(
                "transport-timeout",
                f"viewer did not accept the connection within "
                f"{int(self._connect_timeout * 1000)}ms; staying detached",
            )
            return
        except Exception as exc:
            self._state = "idle"
            self._warner.warn("transport-connect", "failed to open WebSocket", exc)
            return

        was_attached = False
        try:
            await connection.send(self._hooks.build_hello())
            ack = await asyncio.wait_for(
                self._await_ack(connection), timeout=self._handshake_timeout
            )
            if ack is None:
                return
            self._outbox.clear()
            self._state = "attached"
            was_attached = True
            self._hooks.on_attached(ack)
            await self._pump(connection)
        except asyncio.CancelledError:
            raise
        except asyncio.TimeoutError:
            self._warner.warn(
                "transport-ack-timeout",
                f"viewer did not complete the handshake within "
                f"{int(self._handshake_timeout * 1000)}ms; staying detached",
            )
        except Exception as exc:
            self._warner.warn("transport-io", "viewer connection lost", exc)
        finally:
            if self._state != "disposed":
                self._state = "idle"
            self._outbox.clear()
            try:
                await connection.close()
            except Exception:
                pass
            if was_attached:
                # FAIL-OPEN: no debugger, no holds.
                try:
                    self._hooks.on_detached()
                except Exception:
                    pass

    async def _await_ack(self, connection: Any) -> Optional[Dict[str, Any]]:
        """Read frames until ``hello.ack``; anything before it is ignored."""
        while True:
            try:
                raw = await connection.recv()
            except asyncio.CancelledError:
                raise
            except Exception:
                return None
            result = parse_envelope_json(_to_text(raw))
            if result.kind == "version-mismatch":
                self._warner.warn(
                    "transport-version",
                    f"viewer speaks protocol v{result.received}, this client speaks "
                    f"v{PROTOCOL_VERSION}; staying detached",
                )
                return None
            if result.kind != "ok" or result.envelope is None:
                continue
            envelope = result.envelope
            if envelope.get("type") != "hello.ack":
                continue
            payload = envelope.get("payload") or {}
            versions = payload.get("versions") or {}
            if versions.get("protocol") != PROTOCOL_VERSION:
                self._warner.warn(
                    "transport-version",
                    f"viewer acked protocol v{versions.get('protocol')}, this client "
                    f"speaks v{PROTOCOL_VERSION}; staying detached",
                )
                return None
            return payload

    async def _pump(self, connection: Any) -> None:
        """Read control frames while a writer task drains the outbox."""
        writer = asyncio.get_event_loop().create_task(self._write_loop(connection))
        try:
            async for raw in connection:
                if self._state != "attached":
                    break
                result = parse_envelope_json(_to_text(raw))
                if result.kind == "ok" and result.envelope is not None:
                    self._hooks.on_control(result.envelope)
                elif result.kind == "invalid":
                    self._warner.warn(
                        "transport-invalid", f"ignoring invalid frame: {result.reason}"
                    )
                # `unknown-type` and `version-mismatch` mid-stream: tolerate.
        finally:
            writer.cancel()
            try:
                await writer
            except (asyncio.CancelledError, Exception):
                pass

    async def _write_loop(self, connection: Any) -> None:
        wake = self._wake
        while True:
            if not self._outbox:
                if wake is None:
                    await asyncio.sleep(0.01)
                    continue
                await wake.wait()
                wake.clear()
                continue
            try:
                frame = self._outbox.popleft()
            except IndexError:  # pragma: no cover - drained concurrently
                continue
            await connection.send(frame)


def _to_text(raw: Any) -> str:
    if isinstance(raw, str):
        return raw
    if isinstance(raw, (bytes, bytearray, memoryview)):
        try:
            return bytes(raw).decode("utf-8", "replace")
        except Exception:  # pragma: no cover
            return ""
    return str(raw)

File: ruby/lib/graphmind/transport.rb
# frozen_string_literal: true

require_relative "protocol"
require_relative "websocket"

module Graphmind
  # WebSocket transport: lazy connect with a hard connect timeout, the
  # hello/hello.ack handshake, background reconnect, and fail-open detach.
  #
  # Port of packages/client/src/transport.ts. It never raises into callers:
  # every failure degrades to "detached" plus a rate-limited warning, and a
  # background retry keeps trying.
  #
  # Threading: one supervisor thread owns the socket and the read loop; while
  # attached it spawns one writer thread that drains the outbox, so a host
  # thread calling #enqueue never blocks on the network. Both are ordinary
  # (non-main) threads, so they cannot keep a Ruby process alive at exit.
  #
  # Reconnect timing has two regimes, because they are not the same problem:
  #   * never attached (no debugger running) — retry every retry_interval;
  #   * lost an established attachment (a blip) — the debugger existed
  #     milliseconds ago and every millisecond dark eats the ring buffer, so
  #     burst first (FAST_RETRY_STEPS) and then settle.
  class Transport
    FAST_RETRY_STEPS = [0.2, 0.4, 0.8].freeze

    # Hard cap on frames waiting for a slow socket. Oldest are dropped first;
    # the ring buffer, not this queue, is what guarantees replay-on-attach.
    MAX_OUTBOX = 10_000

    Hooks = Struct.new(:build_hello, :on_attached, :on_detached, :on_control, keyword_init: true)

    attr_reader :url

    def initialize(url:, hooks:, warner:, connect_timeout: 0.3, handshake_timeout: 1.0,
                   retry_interval: 10.0)
      @url = url
      @hooks = hooks
      @warner = warner
      @connect_timeout = connect_timeout
      @handshake_timeout = handshake_timeout
      @retry_interval = retry_interval

      @state = :idle # :idle | :connecting | :attached | :disposed
      @started = false
      @mutex = Mutex.new
      @wake = ConditionVariable.new
      @kicked = false
      @supervisor = nil
      @connection = nil
      @outbox = nil
      @writer = nil
      @fast_retry_step = FAST_RETRY_STEPS.length
    end

    def attached? = @state == :attached
    def state = @state

    # Begin connecting (idempotent). Called lazily on first session use.
    def start
      return if @started || @state == :disposed

      @started = true
      @supervisor = Thread.new { supervise }
      @supervisor.name = "graphmind-transport" if @supervisor.respond_to?(:name=)
      @supervisor.report_on_exception = false
      nil
    end

    # Force an attempt now instead of waiting out the retry interval.
    def kick
      return if @state == :disposed
      return start unless @started

      @mutex.synchronize do
        @kicked = true
        @wake.broadcast
      end
      nil
    end

    # Queue one serialized frame. Non-blocking, safe from any thread.
    def enqueue(frame)
      return false unless @state == :attached

      outbox = @outbox
      return false if outbox.nil?

      # Drop-oldest so a stalled socket cannot grow without bound.
      begin
        outbox.pop(true) while outbox.size >= MAX_OUTBOX
      rescue ThreadError
        nil
      end
      outbox << frame
      true
    rescue StandardError
      false
    end

    # Approximate number of frames still waiting to be written.
    def pending = (@outbox&.size || 0)

    def dispose
      return if @state == :disposed

      @state = :disposed
      teardown_connection
      @mutex.synchronize { @wake.broadcast }
      supervisor = @supervisor
      @supervisor = nil
      supervisor&.join(0.5)
      supervisor&.kill if supervisor&.alive?
      nil
    end

    private

    def supervise
      until @state == :disposed
        @mutex.synchronize { @kicked = false }
        begin
          attempt
        rescue StandardError => e
          @warner.warn("transport-loop", "transport loop error", e)
        end
        break if @state == :disposed

        wait_for_retry
      end
    rescue StandardError => e
      @warner.warn("transport-loop", "transport supervisor stopped", e)
    end

    def wait_for_retry
      delay = next_retry_delay
      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + delay
      @mutex.synchronize do
        loop do
          break if @kicked || @state == :disposed

          remaining = deadline - Process.clock_gettime(Process::CLOCK_MONOTONIC)
          break if remaining <= 0

          @wake.wait(@mutex, remaining)
        end
        @kicked = false
      end
    end

    def next_retry_delay
      step = FAST_RETRY_STEPS[@fast_retry_step]
      return @retry_interval if step.nil?

      @fast_retry_step += 1
      step < @retry_interval ? step : @retry_interval
    end

    def attempt
      return if @state == :disposed

      @state = :connecting
      connection =
        begin
          WebSocket.connect(@url, connect_timeout: @connect_timeout,
                                  handshake_timeout: @handshake_timeout)
        rescue WebSocket::Error => e
          @state = :idle
          @warner.warn("transport-connect", "could not reach the GraphMind viewer; staying detached", e)
          return
        rescue StandardError => e
          @state = :idle
          @warner.warn("transport-connect", "failed to open the viewer WebSocket", e)
          return
        end

      was_attached = false
      begin
        connection.send_text(@hooks.build_hello.call)
        ack = await_ack(connection)
        return if ack.nil?

        @connection = connection
        @outbox = Thread::Queue.new
        @state = :attached
        was_attached = true
        start_writer(connection, @outbox)
        safely { @hooks.on_attached.call(ack) }
        read_loop(connection)
      rescue WebSocket::Error => e
        @warner.warn("transport-io", "viewer connection lost", e)
      rescue StandardError => e
        @warner.warn("transport-io", "viewer connection error", e)
      ensure
        @state = :idle unless @state == :disposed
        stop_writer
        @connection = nil
        @outbox = nil
        begin
          connection.abort!
        rescue StandardError
          nil
        end
        if was_attached
          # Losing a live attachment arms the fast-reconnect burst.
          @fast_retry_step = 0
          # FAIL-OPEN: no debugger, no holds.
          safely { @hooks.on_detached.call }
        end
      end
    end

    # Read frames until hello.ack, honouring the handshake budget. Returns the
    # ack payload, or nil when the connection must be abandoned.
    def await_ack(connection)
      deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + @handshake_timeout
      loop do
        message = connection.read_message(deadline: deadline)
        if message.nil?
          @warner.warn(
            "transport-ack-timeout",
            "viewer did not complete the handshake within #{(@handshake_timeout * 1000).round}ms; staying detached"
          )
          return nil
        end
        return nil if message == :closed

        result = Protocol.parse_envelope_json(message)
        case result.kind
        when :version_mismatch
          @warner.warn(
            "transport-version",
            "viewer speaks protocol v#{result.received}, this client speaks " \
            "v#{Protocol::PROTOCOL_VERSION}; staying detached"
          )
          return nil
        when :ok
          envelope = result.envelope
          next unless envelope["type"] == "hello.ack"

          payload = envelope["payload"] || {}
          versions = payload["versions"] || {}
          if versions["protocol"] != Protocol::PROTOCOL_VERSION
            @warner.warn(
              "transport-version",
              "viewer acked protocol v#{versions['protocol']}, this client speaks " \
              "v#{Protocol::PROTOCOL_VERSION}; staying detached"
            )
            return nil
          end
          return payload
        end
        # invalid / unknown-type before the ack: ignore and keep reading.
      end
    end

    def read_loop(connection)
      while @state == :attached
        message = connection.read_message
        break if message.nil? || message == :closed

        result = Protocol.parse_envelope_json(message)
        case result.kind
        when :ok
          safely { @hooks.on_control.call(result.envelope) }
        when :invalid
          @warner.warn("transport-invalid", "ignoring invalid frame: #{result.reason}")
        end
        # unknown-type / version-mismatch mid-stream: tolerate silently.
      end
    end

    def start_writer(connection, outbox)
      @writer = Thread.new do
        loop do
          frame = outbox.pop
          break if frame.nil? || frame.equal?(STOP)

          begin
            connection.send_text(frame)
          rescue StandardError
            # The socket is gone; the read loop will notice and reconnect.
            begin
              connection.abort!
            rescue StandardError
              nil
            end
            break
          end
        end
      end
      @writer.name = "graphmind-writer" if @writer.respond_to?(:name=)
      @writer.report_on_exception = false
    end

    def stop_writer
      writer = @writer
      @writer = nil
      return if writer.nil?

      begin
        @outbox&.<<(STOP)
      rescue StandardError
        nil
      end
      writer.join(0.2)
      writer.kill if writer.alive?
    end

    def teardown_connection
      connection = @connection
      @connection = nil
      return if connection.nil?

      begin
        connection.abort!
      rescue StandardError
        nil
      end
    end

    def safely
      yield
    rescue StandardError => e
      @warner.warn("transport-hook", "internal error in a transport callback", e)
    end

    STOP = Object.new
    private_constant :STOP
  end
end
